src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -588,77 +588,115 @@ ucs_status_t uct_dc_mlx5_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov,
     return UCS_INPROGRESS;
 }
 
-ucs_status_t uct_dc_mlx5_ep_flush(uct_ep_h tl_ep, unsigned flags,
-                                  uct_completion_t *comp)
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_dc_mlx5_ep_flush_cancel(uct_dc_mlx5_ep_t *ep, uint16_t *sn_p)
 {
-    uct_dc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_dc_mlx5_iface_t);
-    uct_dc_mlx5_ep_t    *ep    = ucs_derived_of(tl_ep, uct_dc_mlx5_ep_t);
-    uct_ib_mlx5_md_t    *md    = ucs_derived_of(iface->super.super.super.super.md,
-                                                uct_ib_mlx5_md_t);
-    uint8_t pool_index         = uct_dc_mlx5_ep_pool_index(ep);
-    ucs_status_t        status;
-    uint16_t            sn;
+    uct_dc_mlx5_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                                uct_dc_mlx5_iface_t);
+    uct_ib_mlx5_md_t *md       =
+            ucs_derived_of(iface->super.super.super.super.md,
+                           uct_ib_mlx5_md_t);
+    ucs_status_t status;
     UCT_DC_MLX5_TXQP_DECL(txqp, txwq);
 
-    if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
-        if (uct_dc_mlx5_iface_dci_can_alloc(iface, pool_index)) {
-            UCT_TL_EP_STAT_FLUSH(&ep->super); /* no sends */
-            return UCS_OK;
-        }
-
-        return UCS_ERR_NO_RESOURCE; /* waiting for dci */
+    if (uct_dc_mlx5_iface_is_dci_rand(iface)) {
+        return UCS_ERR_UNSUPPORTED;
     }
 
-    if (!uct_dc_mlx5_iface_has_tx_resources(iface)) {
-        return UCS_ERR_NO_RESOURCE;
+    if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
+        UCT_TL_EP_STAT_FLUSH(&ep->super);
+        return UCS_OK; /* nothing to cancel */
     }
 
-    if (!uct_dc_mlx5_iface_dci_ep_can_send(ep)) {
+    if (!uct_dc_mlx5_iface_has_tx_resources(iface) ||
+        !uct_dc_mlx5_iface_dci_has_tx_resources(iface, ep->dci)) {
         return UCS_ERR_NO_RESOURCE; /* cannot send */
     }
 
+    ucs_assert(ep->dci != UCT_DC_MLX5_EP_NO_DCI);
+
     status = uct_dc_mlx5_iface_flush_dci(iface, ep->dci);
     if (status == UCS_OK) {
         UCT_TL_EP_STAT_FLUSH(&ep->super);
         return UCS_OK; /* all sends completed */
     }
 
     ucs_assert(status == UCS_INPROGRESS);
-    ucs_assert(ep->dci != UCT_DC_MLX5_EP_NO_DCI);
 
     UCT_DC_MLX5_IFACE_TXQP_GET(iface, ep, txqp, txwq);
-    sn = txwq->sig_pi;
+    *sn_p = txwq->sig_pi;
 
-    if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
-        UCT_DC_MLX5_CHECK_RES(iface, ep);
+    if (ep->flags & UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL) {
+        return UCS_INPROGRESS;
+    }
```

Comment:
maybe worth moving to the top of the func?

Reply:
not sure, if no resources better to return UCS_NO_RESOURCES instead of IN_PROGRESS

Reply:
but, for the consecutive flushes you do not need any resources. It is enough that you already have one which is ongoing

Reply:
> but, for the consecutive flushes you do not need any resources. It is enough that you already have one which is ongoing

but what if the previous flush(CANCEL) was completed? we should start the new one and wait for CQE with error for NOP operation

Reply:
maybe we need to clear the cancel flag then? Also do we really need to allow consecutive flush(CANCEL) ops (as we declaring that nothing has to be posted after the flush cancel)?

Reply:
> maybe we need to clear the cancel flag then? Also do we really need to allow consecutive flush(CANCEL) ops (as we declaring that nothing has to be posted after the flush cancel)?

we allow posting no-op for that reason.
actually, there is the UCT test which checks multiple `flush(CANCEL)`

Reply:
but why do we need it? Can we assume that once flush(CANCEL) completes, all other flush(CANCEL) ops do nothing? Because nothing is allowed to post after the very first flush(CANCEL)

Reply:
>  Can we assume that once flush(CANCEL) completes, all other flush(CANCEL) ops do nothing?

it will require adding another flag which indicates whether flush(CANCEL) was completed or not.
also, we can't change API to disallow multiple flush(CANCEL)

Reply:
can we just clear `UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL` when it is completed? So if any other is starting it will set this flag again. Or we can assume that nothing were posted after the first flush cancel and just return `UCS_OK`

Reply:
> can we just clear UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL when it is completed? So if any other is starting it will set this flag again.

unfortunately, clearing this flag could be problematic, it could lead that UCT user error handler will be invoked twice, see
https://github.com/openucx/ucx/blob/c6ff65954138fbcd5155a202dbe8c07ddefa906e/src/uct/ib/dc/dc_mlx5.c#L1590

> Or we can assume that nothing were posted after the first flush cancel and just return UCS_OK

do you mean always return `UCS_OK` when this flag has already been set by the first `flush(CANCEL)` invocation?
this test will be broken then. it expects to get `UCS_INPROGRESS` when calling flush(CANCEL) two times consequentially.

Reply:
i think we may reconsider this test (and its requirement)

Reply:
> i think we may reconsider this test (and its requirement)

I agree that such strict requirements for flush(CANCEL) aren't actually needed. And it will be great to simplify it as much as we can. Normally, users (e.g. UCP) invoke flush(CANCEL) only once.

Reply:
if we return INPROGRESS here we could post another comp with same sn as the flush that is already inflight, and it will be completed as well when we progress. so i think we can move it to top

Reply:
> if we return INPROGRESS here we could post another comp with same sn as the flush that is already inflight, and it will be completed as well when we progress. so i think we can move it to top

what happens if the previous `flush(CANCEL)` has already been completed? I guess it will lead to posting flush(CANCEL) which never complete, since there is no posted operation to complete it

Reply:
let's discuss this with some quorum, i feel something is missing here